### PR TITLE
fix(agents-md): security hardening dropped from #5087 merge

### PR DESCRIPTION
**Pull request by senamakel**
## Summary

- Lands two security fixes authored during review of #5087 that were pushed to the PR branch but missed the merge: GitHub's cross-repo head propagation lagged, and the PR was merged at the previous head, stranding these commits.
- `fix(agents-md): reject symlinked/escaping AGENTS.md before reading` — addresses the codex P1 review thread on #5087.
- `fix(agents-md): redact system-prompt debug log when AGENTS.md present` — addresses the codex P2 review thread on #5087.

## Problem

- A project-controlled `AGENTS.md` could be a symlink pointing at a secret outside the root; the loader would read it and the content would flow into the system prompt (and on to the inference provider).
- The full-body system-prompt debug log could print raw AGENTS.md content verbatim when PROFILE/MEMORY were omitted.
- Both fixes had approved/resolved review threads on #5087 but are absent from `main` because the merge happened before the head ref propagated.

## Solution

- Loader refuses symlinked `AGENTS.md` and validates the canonical path stays under its root (`validate_path_within_root`) before reading; Unix regression test added.
- AGENTS.md presence added to the debug-log redaction trigger so prompt bodies containing it are never logged raw.

## Submission Checklist

- [x] Tests added or updated (happy path + at least one failure / edge case) per [Testing Strategy](../gitbooks/developing/testing-strategy.md#failure-path-requirement) — symlink-rejection and escape-path regression tests; 23 agents_md tests pass
- [x] **Diff coverage ≥ 80%** — changed lines covered by the new agents_md tests
- [x] N/A: Coverage matrix updated — no feature rows added/removed/renamed
- [x] N/A: All affected feature IDs from the matrix are listed under `## Related` — no matrix rows affected
- [x] No new external network dependencies introduced (local file handling only)
- [x] N/A: Manual smoke checklist — does not touch release-cut surfaces
- [x] N/A: Linked issue closed via `Closes #NNN` — follow-up to PR #5087 review threads, no standalone issue

## Impact

- Closes a local file-disclosure vector (symlinked AGENTS.md → prompt → provider) and a PII/content leak into debug logs. No behaviour change for regular non-symlinked AGENTS.md files.

## Related

- Closes:
- Follow-up PR(s)/TODOs: none — completes the review remediation of #5087

---

### Linear Issue

- Key: N/A
- URL: N/A

### Commit & Branch

- Branch: fix/agents-md-security-followup
- Commit SHA: 6b97e643b

### Validation Run

- [x] N/A `pnpm --filter openhuman-app format:check` — no frontend changes
- [x] N/A `pnpm typecheck` — no TypeScript changes
- [x] Focused tests: `cargo test --lib agents_md` (23 passed)
- [x] Rust fmt/check (if changed): `GGML_NATIVE=OFF cargo check` clean, `cargo fmt --check` clean
- [x] N/A Tauri fmt/check (if changed): `app/src-tauri` untouched

### Validation Blocked

- `command:` N/A
- `error:` N/A
- `impact:` N/A

### Behavior Changes

- Intended behavior change: symlinked or root-escaping AGENTS.md files are refused; prompts containing AGENTS.md are redacted in debug logs
- User-visible effect: none for normal AGENTS.md usage

### Parity Contract

- Legacy behavior preserved: regular file loading, layering, capping, and gating unchanged
- Guard/fallback/dispatch parity checks: loader is the single shared path for main + subagent layers, so both are covered

### Duplicate / Superseded PR Handling

- Duplicate PR(s): N/A
- Canonical PR: N/A
- Resolution (closed/superseded/updated): these commits exist on the merged #5087 branch head that GitHub failed to include in the merge

## Summary by CodeRabbit

* **Security Enhancements**
  * Improved protection of sensitive prompt content in debug logs, including instructions from `AGENTS.md`.
  * Prevented loading `AGENTS.md` files that are symlinks or located outside the permitted directory.
  * Added safeguards for invalid or unreadable files while preserving size limits and truncation behavior.

* **Bug Fixes**
  * Sensitive prompt details are now consistently redacted when applicable.

### CODE REVIEW

File: src/openhuman/agent/prompts/agents_md.rs [hunk 1]
```diff
@@ -68,24 +68,64 @@ impl AgentsMdContent {
...
+    let file = match std::fs::File::open(&path) {
+        Ok(f) => f,
+        Err(e) => {
+            // Existence + type were already checked above, so this is a genuine
+            // read/permission failure (or a race). Skip silently like any other
+            // unreadable file.
+            log::debug!("[agents_md] failed to open {}: {e}", path.display());
+            return None;
+        }
+    };
```
[the remark this answers is not in the document]
Comment (senamakel) [reply src/openhuman/agent/prompts/agents_md.rs]: **Addressed** in d0e3337b6. Replaced the stat-then-`File::open`-by-path sequence with a race-free open: `open_no_follow` uses `O_NOFOLLOW | O_NONBLOCK` (Unix) so the kernel **atomically refuses** a final-component symlink at open time — there is no longer any check-to-open window to race, since we never stat-then-open by path. After opening we fstat the *opened handle* and require a regular file (rejecting a directory / FIFO / socket / device a racing writer could substitute), and `O_NONBLOCK` ensures a substituted FIFO/device cannot block the open. `validate_path_within_root` remains as defence in depth against a symlinked *parent* component. Non-Unix keeps a best-effort pre-check (Windows symlink creation needs elevation). Added `fifo_agents_md_is_refused_without_hanging` (a FIFO is exactly the non-regular file a racing writer would swap in) and the existing `symlinked_agents_md_is_refused` now exercises the race-free `O_NOFOLLOW` path. A deterministic swap-between-check-and-open test is omitted deliberately: with `O_NOFOLLOW` there is no check-then-open step to interpose on, so the only way to exercise it would be an injected fs seam that would not reflect the real code path.
